Worker crash when PS replica is set to 0

Feedback from QA. Should we make this case a collective mode or restrict the replica to be positive if PS is specified?
![83e4844f7dedc90edce69df83248786e](https://user-images.githubusercontent.com/31800808/114494360-d16e9700-9c4e-11eb-99de-6f5962f58f2d.PNG)

Tried to reproduce but got a different error.
I changed the PS replica to 0 in <https://github.com/PaddleFlow/paddle-operator/blob/main/deploy/examples/wide_and_deep.yaml> and applied the job, then the controller crashed.
```
$ make run
/home/zzy/go/src/github.com/Ruminateer/paddle-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
/home/zzy/go/src/github.com/Ruminateer/paddle-operator/bin/controller-gen "crd:maxDescLen=0,trivialVersions=true,preserveUnknownFields=false" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
go run ./main.go
2021-04-13T14:36:01.468+0800    INFO    controller-runtime.metrics      metrics server is starting to listen    {"addr": ":8080"}
2021-04-13T14:36:01.471+0800    INFO    setup   starting manager
2021-04-13T14:36:01.572+0800    INFO    controller-runtime.manager      starting metrics server {"path": "/metrics"}
2021-04-13T14:36:01.572+0800    INFO    controller-runtime.manager.controller.paddlejob Starting EventSource    {"reconciler group": "batch.paddlepaddle.org", "reconciler kind": "PaddleJob", "source": "kind source: /, Kind="}
2021-04-13T14:36:01.672+0800    INFO    controller-runtime.manager.controller.paddlejob Starting EventSource    {"reconciler group": "batch.paddlepaddle.org", "reconciler kind": "PaddleJob", "source": "kind source: /, Kind="}
2021-04-13T14:36:01.672+0800    INFO    controller-runtime.manager.controller.paddlejob Starting EventSource    {"reconciler group": "batch.paddlepaddle.org", "reconciler kind": "PaddleJob", "source": "kind source: /, Kind="}
2021-04-13T14:36:01.672+0800    INFO    controller-runtime.manager.controller.paddlejob Starting EventSource    {"reconciler group": "batch.paddlepaddle.org", "reconciler kind": "PaddleJob", "source": "kind source: /, Kind="}
2021-04-13T14:36:01.673+0800    INFO    controller-runtime.manager.controller.paddlejob Starting Controller     {"reconciler group": "batch.paddlepaddle.org", "reconciler kind": "PaddleJob"}
2021-04-13T14:36:01.673+0800    INFO    controller-runtime.manager.controller.paddlejob Starting workers        {"reconciler group": "batch.paddlepaddle.org", "reconciler kind": "PaddleJob", "worker count": 1}
2021-04-13T14:39:47.848+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5261"}
2021-04-13T14:39:47.915+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5263"}
2021-04-13T14:39:47.916+0800    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"PaddleJob","namespace":"default","name":"wide-ande-deep","uid":"d9ba546a-9341-4c73-b66a-170593a80b8d","apiVersion":"batch.paddlepaddle.org/v1","resourceVersion":"5263"}, "reason": "Created", "message": "created  wide-ande-deep-worker-0"}
2021-04-13T14:39:47.979+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5267"}
2021-04-13T14:39:47.979+0800    DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"PaddleJob","namespace":"default","name":"wide-ande-deep","uid":"d9ba546a-9341-4c73-b66a-170593a80b8d","apiVersion":"batch.paddlepaddle.org/v1","resourceVersion":"5267"}, "reason": "Created", "message": "created  wide-ande-deep-worker-1"}
2021-04-13T14:39:48.006+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5273"}
2021-04-13T14:39:48.017+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5273"}
2021-04-13T14:39:48.029+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5275"}
2021-04-13T14:39:48.041+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5277"}
2021-04-13T14:39:48.061+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5278"}
2021-04-13T14:39:48.072+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5279"}
2021-04-13T14:39:48.092+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5279"}
2021-04-13T14:39:48.172+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5279"}
2021-04-13T14:39:48.333+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5279"}
2021-04-13T14:39:48.653+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5279"}
2021-04-13T14:39:48.669+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5288"}
2021-04-13T14:39:48.686+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5289"}
2021-04-13T14:39:49.007+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5289"}
2021-04-13T14:39:49.026+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5292"}
2021-04-13T14:39:49.045+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5292"}
2021-04-13T14:39:49.054+0800    INFO    controllers.PaddleJob   Reconcile       {"paddlejob": "default/wide-ande-deep", "version": "5294"}
E0413 14:39:49.061898   14502 runtime.go:78] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 127 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x16b6a80, 0xc0008fa5a0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:48 +0x82
panic(0x16b6a80, 0xc0008fa5a0)
        /usr/lib/go-1.13/src/runtime/panic.go:679 +0x1b2
github.com/paddleflow/paddle-operator/controllers.constructConfigMap(0xc00090c000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/zzy/go/src/github.com/Ruminateer/paddle-operator/controllers/paddlejob_helper.go:133 +0x136b
github.com/paddleflow/paddle-operator/controllers.(*PaddleJobReconciler).Reconcile(0xc00033ac40, 0x19f0c00, 0xc0006235c0, 0xc000321539, 0x7, 0xc000321520, 0xe, 0xc0006235c0, 0x0, 0x0, ...)
        /home/zzy/go/src/github.com/Ruminateer/paddle-operator/controllers/paddlejob_controller.go:231 +0xd9e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000794140, 0x19f0b40, 0xc00033a180, 0x162b620, 0xc000620dc0)
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x2ea
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000794140, 0x19f0b40, 0xc00033a180, 0x0)
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x22b
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x19f0b40, 0xc00033a180)
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00047c750)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000789f50, 0x19b4e00, 0xc0007ba1b0, 0xc00033a101, 0xc0000ae3c0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00047c750, 0x3b9aca00, 0x0, 0x1, 0xc0000ae3c0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0xaa
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x19f0b40, 0xc00033a180, 0xc00041a8f0, 0x3b9aca00, 0x0, 0x1)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x19f0b40, 0xc00033a180, 0xc00041a8f0, 0x3b9aca00)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x516
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 127 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:55 +0x105
panic(0x16b6a80, 0xc0008fa5a0)
        /usr/lib/go-1.13/src/runtime/panic.go:679 +0x1b2
github.com/paddleflow/paddle-operator/controllers.constructConfigMap(0xc00090c000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/zzy/go/src/github.com/Ruminateer/paddle-operator/controllers/paddlejob_helper.go:133 +0x136b
github.com/paddleflow/paddle-operator/controllers.(*PaddleJobReconciler).Reconcile(0xc00033ac40, 0x19f0c00, 0xc0006235c0, 0xc000321539, 0x7, 0xc000321520, 0xe, 0xc0006235c0, 0x0, 0x0, ...)
        /home/zzy/go/src/github.com/Ruminateer/paddle-operator/controllers/paddlejob_controller.go:231 +0xd9e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000794140, 0x19f0b40, 0xc00033a180, 0x162b620, 0xc000620dc0)
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x2ea
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000794140, 0x19f0b40, 0xc00033a180, 0x0)
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x22b
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x19f0b40, 0xc00033a180)
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00047c750)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000789f50, 0x19b4e00, 0xc0007ba1b0, 0xc00033a101, 0xc0000ae3c0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00047c750, 0x3b9aca00, 0x0, 0x1, 0xc0000ae3c0)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0xaa
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x19f0b40, 0xc00033a180, 0xc00041a8f0, 0x3b9aca00, 0x0, 0x1)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x19f0b40, 0xc00033a180, 0xc00041a8f0, 0x3b9aca00)
        /home/zzy/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /home/zzy/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x516
exit status 2
make: *** [Makefile:29: run] Error 1
```

--- a/controllers/paddlejob_helper.go
+++ b/controllers/paddlejob_helper.go
@@ -127,7 +127,8 @@ func constructConfigMap(pdj *pdv1.PaddleJob, childPods corev1.PodList) (cm *core
 			"PADDLE_TRAINERS":                   strings.Join(workerHosts, ","),
 		},
 	}
-	if pdj.Spec.WithGloo > 0 && pdj.Spec.Intranet != pdv1.Service {
+	// maybe better to use validation webhook to ignore withGloo in collective mode?
+	if pdj.Spec.WithGloo > 0 && pdj.Spec.Intranet != pdv1.Service && len(pservers) > 0 {
 		cm.Data["PADDLE_WITH_GLOO"] = fmt.Sprintf("%d", pdj.Spec.WithGloo)
 		cm.Data["PADDLE_GLOO_RENDEZVOUS"] = "3"
 		cm.Data["PADDLE_GLOO_HTTP_ENDPOINT"] = strings.Replace(pservers[0],

